src/Microsoft.ML.Data/Training/TrainContext.cs

```diff
@@ -0,0 +1,144 @@
+
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Runtime.Training
+{
+    /// <summary>
+    /// A training context is an object instantiable by a user to do various tasks relating to a particular
+    /// "area" of machine learning. A subclass would represent a particular task in machine learning. The idea
+    /// is that a user can instantiate that particular area, and get trainers and evaluators.
+    /// </summary>
+    public abstract class TrainContextBase
+    {
+        protected readonly IHost Host;
+        internal IHostEnvironment Environment => Host;
+
+        protected TrainContextBase(IHostEnvironment env, string registrationName)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckNonEmpty(registrationName, nameof(registrationName));
+            Host = env.Register(registrationName);
+        }
+
+        /// <summary>
+        /// Subclasses of <see cref="TrainContext"/> will provide little "extension method" hookable objects
+        /// (e.g., something like <see cref="BinaryClassificationContext.Trainers"/>). User code will only
+        /// interact with these objects by invoking the extension methods. The actual component code can work
+        /// through <see cref="TrainContextComponentUtils"/> to get more "hidden" information from this object,
+        /// e.g., the environment.
+        /// </summary>
+        public abstract class ContextInstantiatorBase
+        {
+            internal TrainContextBase Owner { get; }
+
+            protected ContextInstantiatorBase(TrainContextBase ctx)
+            {
+                Owner = ctx;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Utilities for component authors that want to be able to instantiate components using these context
+    /// objects. These utilities are not hidden from non-component authoring users per see, but are at least
+    /// registered somewhat less obvious so that they are not confused by the presence.
+    /// </summary>
+    /// <seealso cref="TrainContextBase"/>
+    public static class TrainContextComponentUtils
+    {
+        /// <summary>
+        /// Gets the environment hidden within the instantiator's context.
+        /// </summary>
+        /// <param name="obj">The extension method hook object for a context.</param>
+        /// <returns>An environment that can be used when instantiating components.</returns>
+        public static IHostEnvironment GetEnvironment(TrainContextBase.ContextInstantiatorBase obj)
+        {
+            Contracts.CheckValue(obj, nameof(obj));
+            return obj.Owner.Environment;
+        }
+
+        /// <summary>
+        /// Gets the environment hidden within the context.
+        /// </summary>
+        /// <param name="ctx">The context.</param>
+        /// <returns>An environment that can be used when instantiating components.</returns>
+        public static IHostEnvironment GetEnvironment(TrainContextBase ctx)
```

Comment:
I'm not sure I understand the reasoning for hiding the `Environment` on the Context class. I (as a user) need to pass one in on the public constructor. I know you have one. Why hide the property, but make this static method to get it instead?